app/models/gem_typo.rb

```diff
@@ -0,0 +1,41 @@
+require "rubygems/text"
+
+class GemTypo
+  attr_reader :protected_gem
+
+  include Gem::Text
+
+  DOWNLOADS_THRESHOLD = 10_000_000
+  SIZE_THRESHOLD = 4
+
+  def initialize(rubygem_name)
+    @rubygem_name = rubygem_name.downcase
+  end
+
+  def protected_typo?
+    return false if @rubygem_name.size < GemTypo::SIZE_THRESHOLD
+
+    protected_gems.each do |protected_gem|
+      distance = levenshtein_distance(@rubygem_name, protected_gem)
+      if distance <= distance_threshold
+        @protected_gem = protected_gem
+        return true
+      end
+    end
+
+    false
+  end
+
+  private
+
+  def distance_threshold
+    @rubygem_name.size == GemTypo::SIZE_THRESHOLD ? 1 : 2
+  end
+
+  def protected_gems
+    Rubygem.by_downloads
```

Comment:
What about to use something similar to:

```ruby
    Rubygem.joins(:gem_download)
      .where("gem_downloads.count > ?", GemTypo::DOWNLOADS_THRESHOLD)
      .where.not(name: @rubygem_name)
      .pluck(:name)
```

1. it doesn't do unnecessary order done by `by_downloads` scope
2. it filters out `@rubygem_name` on database level

Reply:
Thanks :+1: Updated. Technically, this PR doesn't need this check but it will be needed when we remove `:create` on validation.